libraries/botbuilder/src/teams/teamsSSOTokenExchangeMiddleware.ts

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import * as signInConstants from '../signInConstants';
+import { ExtendedUserTokenProvider } from '../extendedUserTokenProvider';
+import { Middleware } from '../middlewareSet';
+import { Storage, StoreItem } from '../storage';
+import { TurnContext } from '../turnContext';
+
+import {
+    ActivityTypes,
+    StatusCodes,
+    TokenExchangeInvokeRequest,
+    TokenExchangeInvokeResponse,
+    TokenResponse,
+} from 'botframework-schema';
+
+function getStorageKey(context: TurnContext): string {
+    const activity = context.activity;
+
+    const channelId = activity.channelId;
+    if (!channelId) {
+        throw new Error('invalid activity-missing channelId');
+    }
```

Comment:
The `TeamsSSOTokenExchangeMiddleware` implicitly supports `signin/tokenExchange` Activities for any channel?

Reply:
cc @EricDahlvang

Reply:
As far as I know, these only come from Teams? 

Other channels wouldn't need this middleware anyway.  The true purpose of the middleware is to ensure duplicate token exchange requests are not processed (only one can be processed).

Reply:
Theoretically, then, we could add a filter specifically for events from the teams channel?

Reply:
Bookmarking with https://github.com/microsoft/botbuilder-js/issues/3574 for future consideration.